### ISSUE TITLE
Make space for bar_label annotations automatically

### ISSUE THREAD

**Issue by mwaskom** [matplotlib/matplotlib#20053]
(Following up on gitter conversation)

### Problem

The new `bar_label` method is very handy, and simplifies a task that is an *extremely common* request on stackoverflow and elsewhere.

The only problem is that, due to some complications about including text artists in autoscaling, the annotation for the tallest bar is often either uncomfortably close to the edge of the plot or extending over it:

```python
ax = plt.gca()
bars = ax.bar([0, 1], [1, 2])
ax.bar_label(bars)
```
![image](https://user-images.githubusercontent.com/315810/115868322-49d11700-a40a-11eb-8aad-b533501d7cb0.png)

```python
ax = plt.gca()
bars = ax.barh([0, 1], [1000, 2000])
ax.bar_label(bars)
```
![image](https://user-images.githubusercontent.com/315810/115868386-5ce3e700-a40a-11eb-9d22-d608f5bb8665.png)

This is frustrating because `bar_label` is very close to being a super handy automatic solution, but using it in practice manually setting axis limits to something reasonable.

### Proposed Solution

For my own use, I have cooked up a little helper function that updates the axes data limits to include the texts objects.

```python
ax = plt.gca()
bars = ax.bar([0, 1], [1, 2])
texts = ax.bar_label(bars)

ax.figure.canvas.draw()
for text in texts:
    text_bbox = text.get_window_extent().corners()
    ax.update_datalim(ax.transData.inverted().transform(text_bbox))

ax.autoscale_view()
```

There was some concern raised on gitter that this approach is not fully general because it doesn't handle situations where, because the numbers are too large (and hence wide) or the font is too big, the annotations can't actually fit in the allotted axes space. Although in that case, my code does not blow up, it just gives a result that looks crazy in a fairly-understandable way.

But I do think that something like this (ideally as the default behavior, but possibly as an option) is needed to make `bar_label` actually useful as an "automatic bar labeling" solution.

### Additional context and prior art

https://github.com/matplotlib/matplotlib/issues/19429 is related, although the proposed solution there (do normal clipping on the annotations) would actually exacerbate this issue

**Comment by timhoffm** [comment]
Indeed, this is preferable to clipping (#19429). I'm not too worried on extremely large annotations. These happen rarely and usually don't make much sense. Anyway they'd need user-action. A clipped information is also not helpful.

**Comment by krabo0om** [comment]
Is there any decision if the proposed solution by @mwaskom (or any other) will be implemented? I consider this feature half broken right now.

**Comment by mwaskom** [comment]
Still an issue
